add collective pallet.

SaMotlagh opened the pull request.

SaMotlagh committed:
add collective pallet.

SaMotlagh committed:
finished setting members for council.

SaMotlagh committed:
store council members to subtensor concilmember storgaemap.

SaMotlagh committed:
add storagemap to store concil members inside subtensor pallet.

SaMotlagh committed:
add a council call.

SaMotlagh committed:
add council_set_max_registrations_per_block for council members.

Commit:
Begin switching to pallet_membership for council mgnmt

Commit:
Add test random collective

Commit:
Add pallet-membership to Cargo.toml

Commit:
Add pallet-membership config

Commit:
EnsureRootOrHalfCouncil -> EnsureRootOrMajorityCouncil

Commit:
Add pallet-membership to chainspec.rs

Commit:
Remove --execution native flag from localnet

Commit:
Enforce specific toolchain for project

Commit:
Allow localnet to discover local validators instead of bootnode

Commit:
Re-add pallet-collective from git over fs

Commit:
Properly insert properties map into chainspec

Commit:
Add SudoRuntimeCall, CouncilOrigin to subtensor::Config

Commit:
Add sudo extrinsic to subtensor

Commit:
Add Sudid event to subtensor

Commit:
Fix testing for subtensor by removing collective, membership from runtime

Commit:
Create EnsureCouncilMajority origin in runtime

Commit:
Merge branch 'main' into sudo_decentralization

Rubberbandits commented:
Latest merge conflict resolution includes upping spec_version to 121

shibshib commented on pallets/subtensor/Cargo.toml:
```diff
@@ -43,6 +46,8 @@ sp-tracing = { git = "https://github.com/paritytech/substrate", default-features
 rand = "0.8"
 sp-core = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.39" }
+# Used for sudo decentralization
+pallet-collective = { version = "4.0.0-dev", default-features = false, git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.39" }
```
Not sure if you need it in both test and production dependencies. You likely only need it in the main dependencies. If you need something for testing and not production then you can keep it only in dev dependencies, afaik.

shibshib commented on runtime/src/lib.rs:
```diff
@@ -320,6 +321,42 @@ impl pallet_sudo::Config for Runtime {
+parameter_types! {
+	pub const CouncilMotionDuration: BlockNumber = 3 * MINUTES;
+	pub const CouncilMaxProposals: u32 = 100;
+	pub const CouncilMaxMembers: u32 = 100;
```
I don't think we need that many council members initially, or for testing, can we limit it to 3 for now?

Commit:
Remove pallet-sudo from runtime, remove pallet-collective from dev-deps

Commit:
CouncilMaxMembers -> 3, CouncilMaxProposals -> 10

Commit:
Remove sudo pallet from construct_runtime!

Commit:
Remove sudo pallet config from runtime

Commit:
Remove vestiges of sudo from chain_spec

Commit:
Add localnet chain-spec for testing

Commit:
Fix tabulation in subtensor::sudo comment

Commit:
Re-add sudo to runtime

Commit:
Ensure root for membership mngmt

Commit:
Return DispatchResult of call from sudo

Commit:
Add sudoUncheckedWeight extrinsic to subtensor

Commit:
CouncilMotionDuration -> 1 DAY

Commit:
Move collective pallet into repository

Commit:
:robot: :butter: Reformat source code

Commit:
Add SenateCollective, SenateMembers to manage senate votes

Commit:
Add VoteOrigin, ProposalOrigin to collective pallet

Commit:
Add more test accounts to localnet_genesis

Commit:
Fix removed finney testnet sudo key

Commit:
Add CanPropose, CanVote, GetVotingMembers traits

Commit:
Add CanPropose, CanVote implementations to runtime for collective

Commit:
Remove threshold argument from collective::propose

Commit:
Add max members for senate

Commit:
Remove threshold argument from collective::benchmarks::propose

Commit:
Add GetVotingMembers implementation for Collective

Commit:
Remove unused pallet-collective in subtensor

Commit:
Remove pallet_collective imports from subtensor

Commit:
Add MemberManagement trait to subtensor, implementation in runtime

Commit:
Add swap_member, members to MemberManagement trait

Commit:
Implement new MemberManagement interface methods

Commit:
Add senate module, do_join_senate, do_leave_senate extrinsics to subtensor

Commit:
Add max_members to MemberManagement, fix vec includes for interface

Commit:
Add subtensor::SenateRequiredStakePercentage storage

Commit:
Add benchmarks for join_senate, leave_senate

Commit:
Fix all broken benchmarks

Commit:
Reverse sort stake comparison in join_senate for descending order

Commit:
Fix all benchmarks properly this time

Commit:
Speed up benchmarking process for subtensor

Commit:
Use a full senate for senate benchmarking

Commit:
Update weights and DB reads for join_senate, leave_senate

Commit:
Remove senate member if required stake drops below threshold

Commit:
Update remove_stake benchmark to include senate removal threshold

Commit:
Update remove_stake weight

Commit:
Fix division by zero case in join_senate

Commit:
Remove cast votes of senate member when leaving

Commit:
CouncilMotionDuration 7200 blocks -> 100 blocks

Commit:
Use coldkey as signing origin for senate actions

Add add_vote to TriumvirateInterface
Update benchmarks for new signing scheme
Fix some tabulation differences in benchmarks.rs

Commit:
Update Cargo.lock for subtensor

Commit:
Add collective, membership pallets to subtensor Cargo.toml

Commit:
Update collective testing to support modifications

Commit:
Add triumvirate+senate to subtensor testing

Commit:
Up spec version 121 -> 122

Commit:
Handle result in CollectiveInterface::remove_votes

Commit:
Begin implementing subtensor-specific senate testing

Commit:
Add verbose errors for senate actions

Commit:
Fix staking removal tests broken by weight updates

Commit:
Add pallet-collective and pallet-membership to subtensor std for testing

Commit:
Fix division by zero in remove_stake for senate removal

Commit:
Add has_voted helper to pallet-collective

Commit:
Add complete test suite for senate activities in subtensor

Commit:
Merge branch 'main' into sudo_decentralization

Commit:
Fix errors from resolved merge conflicts

camfairchild commented on pallets/collective/Cargo.toml:
```diff
@@ -0,0 +1,47 @@
+[package]
+name = "pallet-collective"
+version = "4.0.0-dev"
+authors = ["Parity Technologies <admin@parity.io>"]
```
Should be opentensor? Or at least adding opentensor

Rubberbandits replied:
Probably should add it, it’s partially modified

camfairchild commented on pallets/collective/src/migrations/v4.rs:
```diff
@@ -0,0 +1,148 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Why do we need the migration? Does collective already exist in the chain storage?

Rubberbandits replied:
It doesn’t, leftover from forking the pallet. I believe migrations are only triggered when specified essentially.

camfairchild replied:
> It doesn’t, leftover from forking the pallet. I believe migrations are only triggered when specified essentially.

Definitely should remove then

camfairchild reviewed (commented):
Review in progress, just had some questions 
@Rubberbandits

Commit:
Update Cargo.toml authors and remove collective/migrations

Commit:
Remove migrations testing in pallet-collective
